File: Test/AdventureWorksFunctionalModel/Sales/SalesPerson_Functions.cs







using System;
using System.Linq;
using AW.Types;
using NakedFunctions;

namespace AW.Functions {
    public static class SalesPerson_Functions {
        internal static IContext UpdateSalesPerson(
            SalesPerson original, SalesPerson updated, IContext context) =>
            context.WithUpdated(original, new(updated) { ModifiedDate = context.Now() });

        [MemberOrder(1)]
        public static IContext RecalulateSalesYTD(this SalesPerson sp, IContext context) {
            var startOfYear = new DateTime(DateTime.Now.Year, 1, 1);
            var query = from obj in context.Instances<SalesOrderHeader>()
                        where obj.SalesPerson != null && obj.SalesPerson.BusinessEntityID == sp.BusinessEntityID &&
                              obj.StatusByte == 5 &&
                              obj.OrderDate >= startOfYear
                        select obj;
            var newYTD = query.Any() ? query.Sum(n => n.SubTotal) : 0;

            return UpdateSalesPerson(sp, new(sp) { SalesYTD = newYTD }, context);
        }

        [MemberOrder(2)]
        public static IContext ChangeSalesQuota(
            this SalesPerson sp, decimal newQuota, IContext context) =>
            UpdateSalesPerson(sp, new(sp) { SalesQuota = newQuota }, context)
                .WithNew(new SalesPersonQuotaHistory { SalesPerson = sp, SalesQuota = newQuota, QuotaDate = context.Now() });

        [MemberOrder(1)]
        public static IContext ChangeSalesTerritory(this SalesPerson sp, SalesTerritory newTerritory, IContext context) {
            var newHist = new SalesTerritoryHistory { SalesPerson = sp, SalesTerritory = newTerritory, StartDate = context.Now() };
            var prev = sp.TerritoryHistory.First(n => n.EndDate == null);
            SalesTerritoryHistory uPrev = new(prev) { EndDate = context.Now() };
            SalesPerson uSp = new(sp) { SalesTerritory = newTerritory, ModifiedDate = context.Now() };
            return context.WithNew(newHist).WithUpdated(sp, uSp).WithUpdated(prev, uPrev);
        }
    }
}

File: Test/AdventureWorksLegacyModelVB/Sales/SalesPerson_Functions.vb
Namespace AW.Functions
	Public Module SalesPerson_Functions

		<MemberOrder(1)>
		<System.Runtime.CompilerServices.Extension>
		Public Function RecalulateSalesYTD(ByVal sp As SalesPerson, ByVal context As IContext) As IContext
			Dim startOfYear = New DateTime(DateTime.Now.Year, 1, 1)
			Dim query = From obj In context.Instances(Of SalesOrderHeader)()
						Where obj.SalesPerson IsNot Nothing AndAlso obj.SalesPerson.BusinessEntityID = sp.BusinessEntityID AndAlso obj.StatusByte = 5 AndAlso obj.OrderDate >= startOfYear
						Select obj
			Dim newYTD = If(query.Any(), query.Sum(Function(n) n.SubTotal), 0)

			Return context.WithUpdated(sp, NewDirectCast({
				SalesYTD = newYTD,
				ModifiedDate = context.Now()
			}, sp))
		End Function

		<MemberOrder(2)>
		<System.Runtime.CompilerServices.Extension> _
		Public Function ChangeSalesQuota(ByVal sp As SalesPerson, ByVal newQuota As Decimal, ByVal context As IContext) As IContext
			Return context.WithUpdated(sp, NewDirectCast({
				SalesQuota = newQuota,
				ModifiedDate = context.Now()
			}, sp)).WithNew(New SalesPersonQuotaHistory With {
				.SalesPerson = sp,
				.SalesQuota = newQuota,
				.QuotaDate = context.Now()
			})
		End Function

		<MemberOrder(1)>
		<System.Runtime.CompilerServices.Extension> _
		Public Function ChangeSalesTerritory(ByVal sp As SalesPerson, ByVal newTerritory As SalesTerritory, ByVal context As IContext) As IContext
			Dim newHist = New SalesTerritoryHistory With {
				.SalesPerson = sp,
				.SalesTerritory = newTerritory,
				.StartDate = context.Now()
			}
			Dim prev = sp.TerritoryHistory.First(Function(n) n.EndDate Is Nothing)
			Dim uPrev As New SalesTerritoryHistory(prev)
			If True Then
				EndDate = context.Now()
			End If
			Dim uSp As New SalesPerson(sp)
			If True Then
'INSTANT VB WARNING: An assignment within expression was extracted from the following statement:
'ORIGINAL LINE: SalesTerritory = newTerritory, ModifiedDate = context.Now()
				newTerritory, ModifiedDate = context.Now()
				SalesTerritory = newTerritory, ModifiedDate
			End If
			Return context.WithNew(newHist).WithUpdated(sp, uSp).WithUpdated(prev, uPrev)
		End Function
	End Module

End Namespace